### ISSUE TITLE
Bing Maps - Editable polygon

### ISSUE THREAD

**Issue by dahovey** [infusion-code/angular-maps#34]
I've forked this repository and was hoping to add support for editing polygons with Bing Maps, but I had some questions/issues:

1. When I run `npm install` the following error occurs apparently from linting. After removing `prepublish` script from package.json the error no longer occurs.
> ERROR: src/services/bing/bing-cluster.service.ts[248, 75]: Shadowed name: 's'` a
2. When I run `npm run build` I receive errors: 
> Error at C:/temp/angular-maps/src/models/bing/bing-cluster-layer.ts:113:41: The left-hand side of an 'instanceof' expression must be of type 'any', an object type or a type parameter.
Error at C:/temp/angular-maps/src/models/bing/bing-cluster-layer.ts:130:41: The left-hand side of an 'instanceof' expression must be of type 'any', an object type or a type parameter.
Error at C:/temp/angular-maps/src/models/google/google-marker-clusterer.ts:87:41: The left-hand side of an 'instanceof' expression must be of type 'any', an object type or a type parameter.
Error at C:/temp/angular-maps/src/models/google/google-marker-clusterer.ts:102:41: The left-hand side of an 'instanceof' expression must be of type 'any', an object type or a type parameter.
3. Should I create a new branch for my changes?
4. Is there a date you hope to publish current changes in master branch?

Thanks for your help!

**Comment by thor-schueler** [comment]
Hi David. Making polys editable is certainly a worthwhile effort. Let's get your environment up and running first through. 

Your first issue is strange. It is a linter warning, your are correct. Only s is not shadowed. I can however see why the linter would think that. Since you are doing a fresh install, it is likely installing version 5.7.0 of tslint, while I have been using  5.4.2 or thereabouts. I'll upgrade my linter to see what's what, but in the meanwhile, simple change line 248 of bing-cluster.service.ts to 

const payload: (_s: string, _i: IMarkerIconInfo) => void = (s, i) => {
...

That should do the trick for now. You want to put the prepublish back in as you'll need it ongoingly to check various things. 

Your second issue seems to be a problem with the TS compiler. What version of typescript are you using? You are probably running 2.5.3 or something thereabouts. The last passing build (two days ago) was done on 2.3.2. This used to be ok, but apparently rules have changed. I'll look into that, but int the meanwhile, you can change bing-cluster-layer.ts line 113 to 

if (<any>entity instanceof Marker || <any>entity instanceof BingMarker) {
...
                
ditto the other 3 occurrences. 

Thanks for finding those two. 

Now for the fun stuff: 

Yes, you will probably want to create a branch in your fork, that way you can easily revert. Once you are at a good point in your development, submit a pull request. Depending on the size of the pull request, it might take a few days to review it and incorporate it. 

A few caveats though:

- Think through what you want to do and how prior to implementation. Preferably, let's discuss it first. 
- When implementing new functionality, make sure it works against both Google and Bing. There does not have to be 100% feature congruence, but it should mostly work the same (I generally use plunker to confirm functionality across both platforms). Sadly, we've had not have the time yet to create unit tests. 
- Since you will never want to use any references to Bing or Google in the components themselves, you will probably need to think through the abstraction layer that you need to add, and provide a service model which you then can implement against Bing and Google. If it helps, we can talk through it here as your vision crystallizes. 
- We have a number of commercial products built on this module, so i tend to push back against any breaking changes. If there are any, we have think through them to minimize their impact.  

There are couple of good examples on how to work with the drawing components in Bing at https://www.bing.com/api/maps/sdkrelease/mapcontrol/isdk under layers and visualizations. It's a good starting point. Google has something similar at https://developers.google.com/maps/documentation/javascript/examples/, Drawing on the Map section...

Looking forward to it!

**Comment by dahovey** [comment]
For tslinst error I changed to below. A few lines below this function was another variable named `s`. `const payload: (icon: string, info: IMarkerIconInfo) => void = (icon, info) => {`

For typescript compiler errors what fixed issue was using `<any>` for right side of condition.
`if (entity instanceof Marker || <any>entity instanceof BingMarker) {`

Made those changes to get moving on editable changes. So far what I've done:

bing-map.service.ts: 220
``` typescript
  const map = new Microsoft.Maps.Map(el, o);
            const self = this;
            Microsoft.Maps.loadModule('Microsoft.Maps.DrawingTools', function () {
                self._tools = new Microsoft.Maps.DrawingTools(map);
                self._mapInstance = map;
                self._mapResolver(map);
            });
            return;
```

I modified the constructor of BingPolygon to accept instance of `Microsoft.Maps.DrawingTools`. then `SetEditable` function of BingPolygon becomes:
```typescript
public SetEditable(editable: boolean): void {
        const isChanged = this._isEditable !== editable;
        this._isEditable = editable;

        if (!isChanged) {
            return;
        }

        if (this._isEditable) {
            this._tools.edit(this._polygon);
        } else {
            this._tools.finish();
        }
    }
```

**Implementation concern**

How should consumer be notified that editing is complete, and receive new path/paths? I assume this question is also applicable to Google Maps.

**Possible implementation**

MapPolygonDirective.ts
```typescript
    /**
     * This event is fired when the path/paths is changed after editing is finished.
     *
     * @type {EventEmitter<Array<ILatLong> | Array<Array<ILatLong>>>}
     * @memberof MapPolygonDirective
     */
    @Output() PathChanged: EventEmitter<Array<ILatLong> | Array<Array<ILatLong>>> = new EventEmitter<Array<ILatLong> | Array<Array<ILatLong>>>();
```

Then `AddEventListeners` and `CreateEventObservable` could be leveraged for a custom event that allows for the paths, but these methods to be specific to MouseEvents. Or a custom method could be added to PolygonService specific to this event. Either shouldn't be breaking changes, just adding new emitter to the polygon directive.

With this event consumers have the chance to persist the new paths/modify the original path array. 

For me I also want the ability to cancel the editing process, but my solution will be to keep a copy of the original path on my end then if user cancels editing I can re-set the Path and let change detection set the path for the polygon.

**Comment by dahovey** [comment]
I had some trouble testing my changes on the forked repo. I used `npm link` so that my project points to the local `angular-maps` repo. Possibly since I'm using Angular CLI, I get errors likely because of the typescript and the index.ts present in the repo. My work around was to run `npm pack` then use `npm install <packagepath>` from my project. Do you have another way to test local changes?

How are you using plunkr to test your changes? Is after you've published them to npm?

**Comment by thor-schueler** [comment]
Yes, I usually test locally by creating a cli project, installing the latest published version, then perform a local build and simply copy the contents of the dist folder into the appropriate spot in the npm_modules folder of the cli project. 

The npm link will work, but you will have to modify your tsconfig file to prevent inclusion of the ts sources that are now in your npm_modules folder... I find the copying easy enough ;)

I like raising an event when the path has changed. Very clean. 

I should have a revision to the project soon that corrects the build errors you experienced with the latest version of tslint and typescript.

**Comment by thor-schueler** [comment]
I just pushed a change that should resolve the linting and compile issues you've experienced.

**Comment by dahovey** [comment]
For testing Google Maps locally I don't have an api key. Do I need to go through the process of acquiring one from Google to test changes? Just to see if I get up-and-running I tried the key you are using for plunker but maybe it only works for specific domain name?

The project I'm working on is using Bing, so I have a key to use to test with at least locally.

I created my own repo just for demo. I'll be pushing it to my own account and you can feel free to use it if provides any value for you.

**Comment by dahovey** [comment]
Pull request created. I also created repo for demo/testing. [https://github.com/hoveytechllc/angular-maps-demo](https://github.com/hoveytechllc/angular-maps-demo)

When using demo site you can provide query string `?type=google` or `?type=bing` to register their respective loaders.

**Comment by thor-schueler** [comment]
I was testing some more today and it looks like the addition is breaking late loading polygons (SetPath requires the polygon to be editable) and mutli-path polygons. I'll do some more troubleshooting tomorrow, i hope it will be a quick fix.

**Comment by thor-schueler** [comment]
I worked through the issues I ran into and made some minor modifications. There are still three open items:

1. In Google Maps, editing polys with more than a few dozen points is painfully slow. This is not a blocker, but I'll see whether there are any optimizations we can do. I kinda doubt it though.
2. In Bing Maps, editing complex polys will only put the first ring into edit mode. The other rings are not editable. I think that can possibly be resolved by treating the rings as individual polys during the editing. I will reach out to @rbrundritt to get some advise on how to handle this potentially. Not a blocker though, just a limitation for now. 
3. In Bing Maps, I'd like to improve the way we load the modules before I commit the merge. Prob tomorrow. 

Still need to bring that same approach to polylines, but that will prob after the merge. Looking good...

**Comment by dahovey** [comment]
Sounds good.

On 3, I agree. Loading the module during Map initialization was probably not the most ideal. It would be nice if the `BingMapService` could load the module when needed. `BingPolygon` could just resolve a promise that returns the `DrawingTools` instance.

**Comment by thor-schueler** [comment]
Yes, that is what I am planning to do.

**Comment by rbrundritt** [comment]
For editing complex polygons, currently you will need to treat each ring as it's own polygon for editing. I will see add a feature request to support complex polygon editing.

**Comment by thor-schueler** [comment]
This has been merged with #40 . Will be published in rev 0.7.0.

**Comment by thor-schueler** [comment]
Added samples to the samples page - Editable polygons - [Bing](https://plnkr.co/tYPIUQ) | [Google](https://plnkr.co/9EpFlp)
